src/Symfony/Component/Cache/Adapter/RedisTagAwareAdapter.php

```diff
@@ -0,0 +1,180 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Cache\Adapter;
+
+use Predis;
+use Predis\Connection\Aggregate\ClusterInterface;
+use Predis\Response\Status;
+use Symfony\Component\Cache\Exception\LogicException;
+use Symfony\Component\Cache\Marshaller\MarshallerInterface;
+use Symfony\Component\Cache\Traits\RedisTrait;
+
+/**
+ * Stores tag id <> cache id relationship as a Redis Set, lookup on invalidation using sPOP.
+ *
+ * Set (tag relation info) is stored without expiry (non-volatile), while cache always gets an expiry (volatile) even
+ * if not set by caller. Thus if you configure redis with the right eviction policy you can be safe this tag <> cache
+ * relationship survives eviction (cache cleanup when Redis runs out of memory).
+ *
+ * Requirements:
+ *  - Server: Redis 3.2+
+ *  - Client: PHP Redis 3.1.3+ OR Predis
+ *  - Redis Server(s) configured with any `volatile-*` eviction policy, OR `noeviction` if it will NEVER fill up memory
+ *
+ * Design limitations:
+ *  - Max 2 billion cache keys per cache tag
+ *    E.g. If you use a "all" items tag for expiry instead of clear(), that limits you to 2 billion cache items as well
+ *
+ * @see https://redis.io/topics/lru-cache#eviction-policies Documentation for Redis eviction policies.
+ * @see https://redis.io/topics/data-types#sets Documentation for Redis Set datatype.
+ * @see https://redis.io/commands/spop Documentation for sPOP operation, capable of retriving AND emptying a Set at once.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ * @author André Rømcke <andre.romcke+symfony@gmail.com>
+ *
+ * @experimental in 4.3
+ */
+class RedisTagAwareAdapter extends AbstractTagAwareAdapter
+{
+    use RedisTrait;
+
+    /**
+     * Redis "Set" can hold more than 4 billion members, here we limit ourselves to PHP's > 2 billion max int (32Bit).
+     */
+    private const POP_MAX_LIMIT = 2147483647 - 1;
```

Comment:
It's probably not important, but strictly speaking it should be:

```php
// 2^32 - 1 = max number of members in a Redis set
private const POP_MAX_LIMIT = min(2^32 - 1, PHP_INT_MAX - 1);
```

Reply:
I like it. WDYT @nicolas-grekas ?

Reply:
works for me - but `min()` cannot be part of a const afaik

Reply:
That's right. Maybe this other expression could work?

```php
private const POP_MAX_LIMIT = PHP_INT_SIZE > 4 ? 2^32 - 1 : PHP_INT_MAX - 1
```

Reply:
Indeed, tried to do something like that in the very beginning of this.